Content.Client/Singularity/SingularityOverlay.cs

```diff
@@ -86,6 +88,39 @@ protected override void Draw(in OverlayDrawArgs args)
             worldHandle.DrawRect(args.WorldAABB, Color.White);
             worldHandle.UseShader(null);
         }
+
+        private void OnProjectFromScreenToMap(ref ProjectScreenToMapEvent args)
+        {   // Mostly copypasta from the singularity shader.
+            Vector2 finalCoords = args.ScreenPosition;
+            Vector2 delta;
+            float distance = 0.0f;
+            float deformation = 0.0f;
+            float maxDistance = MaxDistance * EyeManager.PixelsPerMeter;
+    
+            for (int i = 0; i < MaxCount && i < _count; i++)
+            {
+                delta = args.ScreenPosition - _positions[i];
+                distance = (delta / args.ClydeViewport.RenderScale).Length;
+
+                deformation = _intensities[i] / MathF.Pow(distance, _falloffPowers[i]);
+                
+                // ensure deformation goes to zero at max distance
+                // avoids long-range single-pixel shifts that are noticeable when leaving PVS.
+                
+                if (distance >= maxDistance) {
+                    deformation = 0.0f;
+                } else {
+                    deformation *= (1.0f - MathF.Pow(distance/maxDistance, 4.0f));
+                }
+                
+                if(deformation > 0.8)
+                    deformation = MathF.Pow(deformation, 0.3f);
+
+                Vector2 displacement = delta * deformation;
+                finalCoords -= displacement;
+            }
+
+            args.ScreenPosition = finalCoords;
+        }
```

Comment:
Don't hardcode this here, I believe you could compute a transform matrix for this value, so instead have an event requesting interested parties provide a transform for screen space and handle it wherever client singulo is.

Reply:
Are you referring to the entire function or the final assignment of the screen position?

If you're referring to the former I have legitimately no idea how to convert the function into a matrix given that the amount of its scaling is dependent on the distance from the center of the singularity.

If you're referring to the latter I can do it pretty easily.